Crash on converting mixed headers from h1 to h2

## Scope
Configure backend to send headers in mixed order: put header `B` between two headers `A`. e.g.:
 
	location /short_headers {
		add_header X-Extra-Data1 "q";
		add_header X-Extra-Data2 "q";
		add_header X-Extra-Data1 "q";
		root /srv/http ;
	}

The crash will happen once the message is converted from h1 to h2. This may happen in two places:
- saving responses to cache (headers are stored in native h2 format)
- forward to h2 client

The issue doesn't reproduced if h2 request contains mixed headers orders.
Probably related to #1387

## Testing

Currently we have very limited h2 test coverage, only basic GET/POST request. It's possible to extend them, but more sophisticated fussing test is required.

## Logs

Crash on caching:
```
------------[ cut here ]------------
[  680.957333] kernel BUG at /home/user/qtc/release/tempesta/tempesta_db/core/htrie.c:636!
[  680.962458] invalid opcode: 0000 [#1] SMP PTI
[  680.963866] Modules linked in: tempesta_fw(O) tempesta_db(O) sha256_ssse3 sha512_ssse3 sha512_generic ccm tempesta_tls(O) tempesta_lib(O) kvm_intel kvm irqbypass crct10dif_pclmul crc32_pclmul iTCO_wdt iTCO_vendor_support snd_hda_codec_generic snd_hda_intel snd_hda_codec ghash_clmulni_intel snd_hda_core snd_hwdep snd_pcm snd_timer virtio_balloon snd sg soundcore virtio_console evdev serio_raw pcspkr virtio_gpu lpc_ich ttm mfd_core binfmt_misc drm_kms_helper button drm shpchp ip_tables x_tables autofs4 ext4 crc16 mbcache jbd2 fscrypto ecb aesni_intel crypto_simd cryptd glue_helper aes_x86_64 crc32c_generic sr_mod cdrom virtio_blk virtio_net crc32c_intel psmouse ahci libahci libata i2c_i801 ehci_pci virtio_pci sym53c8xx scsi_transport_spi uhci_hcd ehci_hcd scsi_mod virtio_ring virtio usbcore usb_common
[  680.966427] CPU: 0 PID: 550 Comm: nginx Tainted: G           O    4.14.0-tempesta-kmemleak-amd64 #1 Debian 4.14.32-tfw7-1
[  680.966427] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS ?-20191223_100556-anatol 04/01/2014
[  680.966427] task: ffff8a5876510f00 task.stack: ffffb922012e4000
[  680.966427] RIP: 0010:tdb_htrie_extend_rec+0x7c/0x90 [tempesta_db]
[  680.966427] RSP: 0018:ffff8a58e5403518 EFLAGS: 00010246
[  680.966427] RAX: 0000000000000000 RBX: ffff8a58d1405300 RCX: ffff8a58d1405380
[  680.966427] RDX: 0000000000005380 RSI: 0000000000000000 RDI: ffffffffc0742cc6
[  680.966427] RBP: ffff8a58d1400000 R08: 0000000000000000 R09: ffffffffc08112f0
[  680.966427] R10: 0000000000000000 R11: 000000000000002e R12: ffff8a58bcbc4100
[  680.966427] R13: 0000000000000000 R14: 0000000000000006 R15: 0000000000000006
[  680.966427] FS:  00007fb0d64ac740(0000) GS:ffff8a58e5400000(0000) knlGS:0000000000000000
[  680.966427] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  680.966427] CR2: 000055bce7ebf000 CR3: 000000007f40e005 CR4: 00000000003606f0
[  680.966427] Call Trace:
[  680.966427]  <IRQ>
[  680.966427]  __tfw_cache_strcpy.isra.9+0x98/0x140 [tempesta_fw]
[  680.966427]  ? tfw_cache_build_resp_hdr+0x170/0x170 [tempesta_fw]
[  680.966427]  tfw_cache_h2_copy_str+0x77/0xd0 [tempesta_fw]
[  680.966427]  tfw_cache_copy_resp+0xb84/0x1240 [tempesta_fw]
[  680.966427]  ? tdb_alloc_data+0xf7/0x150 [tempesta_db]
[  680.966427]  __cache_add_node+0xb1/0xe0 [tempesta_fw]
[  680.966427]  ? tfw_h2_resp_add_loc_hdrs+0xa0/0xa0 [tempesta_fw]
[  680.966427]  ? tfw_h2_resp_add_loc_hdrs+0xa0/0xa0 [tempesta_fw]
[  680.966427]  tfw_cache_do_action+0x354/0xe20 [tempesta_fw]
[  680.966427]  ? __hdr_name_cmp+0x140/0x320 [tempesta_fw]
[  680.966427]  ? frang_resp_handler+0x162/0x3a0 [tempesta_fw]
[  680.966427]  ? frang_resp_handler+0x162/0x3a0 [tempesta_fw]
[  680.966427]  ? tfw_http_msg_hdr_close+0x1be/0x200 [tempesta_fw]
[  680.966427]  ? tfw_h2_resp_add_loc_hdrs+0xa0/0xa0 [tempesta_fw]
[  680.966427]  tfw_cache_process+0xbc/0x2a0 [tempesta_fw]
[  680.966427]  ? tfw_gfsm_move+0x14b/0x180 [tempesta_fw]
[  680.966427]  ? kvm_clock_get_cycles+0x1a/0x20
[  680.966427]  tfw_http_resp_cache+0xe1/0x150 [tempesta_fw]
[  680.966427]  tfw_http_msg_process_generic+0x893/0xd10 [tempesta_fw]
[  680.966427]  ? eth_header+0x25/0xc0
[  680.966427]  ? pg_skb_alloc+0x253/0x470
[  680.966427]  tfw_http_msg_process+0x84/0xc0 [tempesta_fw]
[  680.966427]  __gfsm_fsm_exec+0x56/0x90 [tempesta_fw]
[  680.966427]  ? consume_skb+0x27/0x80
[  680.966427]  tfw_connection_recv+0x41/0x60 [tempesta_fw]
[  680.966427]  ? tfw_connection_send+0x30/0x30 [tempesta_fw]
[  680.966427]  ss_tcp_process_data+0x1ea/0x480 [tempesta_fw]
[  680.966427]  ? update_load_avg+0x40d/0x580
[  680.966427]  ss_tcp_data_ready+0x43/0x90 [tempesta_fw]
[  680.966427]  tcp_data_queue+0x4f5/0xc50
[  680.966427]  tcp_rcv_established+0x27c/0x570
[  680.966427]  tcp_v4_do_rcv+0x129/0x1d0
[  680.966427]  tcp_v4_rcv+0x947/0xa50
[  680.966427]  ip_local_deliver_finish+0x9a/0x1c0
[  680.966427]  ip_local_deliver+0x6b/0xe0
[  680.966427]  ? nf_hook_slow+0x43/0xc0
[  680.966427]  ip_rcv+0x289/0x3c0
[  680.966427]  ? inet_del_offload+0x40/0x40
[  680.966427]  __netif_receive_skb_core+0x84f/0xb30
[  680.966427]  ? blk_mq_free_request+0xfe/0x110
[  680.966427]  ? process_backlog+0xa3/0x160
[  680.966427]  process_backlog+0xa3/0x160
[  680.966427]  net_rx_action+0x28e/0x3f0
[  680.966427]  __do_softirq+0x10f/0x2a8
[  680.966427]  do_softirq_own_stack+0x2a/0x40
[  680.966427]  </IRQ>
[  680.966427]  do_softirq.part.17+0x3d/0x50
[  680.966427]  __local_bh_enable_ip+0x55/0x60
[  680.966427]  ip_finish_output2+0x1b3/0x3b0
[  680.966427]  ? ip_output+0x71/0xe0
[  680.966427]  ip_output+0x71/0xe0
[  680.966427]  ? ip_queue_xmit+0x5c/0x3b0
[  680.966427]  tcp_transmit_skb+0x535/0x9c0
[  680.966427]  tcp_write_xmit+0x223/0x11c0
[  680.966427]  ? _copy_from_iter_full+0x93/0x230
[  680.966427]  __tcp_push_pending_frames+0x2d/0xd0
[  680.966427]  tcp_sendmsg_locked+0x68e/0xe60
[  680.966427]  tcp_sendmsg+0x27/0x40
[  680.966427]  sock_sendmsg+0x36/0x40
[  680.966427]  sock_write_iter+0x87/0x100
[  680.966427]  do_iter_readv_writev+0x11b/0x190
[  680.966427]  do_iter_write+0x80/0x190
[  680.966427]  vfs_writev+0x98/0x110
[  680.966427]  ? do_writev+0x5e/0xf0
[  680.966427]  do_writev+0x5e/0xf0
[  680.966427]  ? SyS_pread64+0x90/0xb0
[  680.966427]  do_syscall_64+0x8d/0x120
[  680.966427]  entry_SYSCALL_64_after_hwframe+0x3d/0xa2
[  680.966427] RIP: 0033:0x7fb0d4de4fe4
[  680.966427] RSP: 002b:00007ffdc4bb66a8 EFLAGS: 00000246 ORIG_RAX: 0000000000000014
[  680.966427] RAX: ffffffffffffffda RBX: 00007fb0d64802c8 RCX: 00007fb0d4de4fe4
[  680.966427] RDX: 0000000000000002 RSI: 00007ffdc4bb6760 RDI: 0000000000000008
[  680.966427] RBP: 000055bce9ec2ab8 R08: 000055bce9ec2ba0 R09: 00007ffdc4bb6770
[  680.966427] R10: 000055bce9ea4d50 R11: 0000000000000246 R12: 00007ffdc4bb6740
[  680.966427] R13: 000055bce7f0048c R14: 000000007fffefff R15: 0000000000000157
[  680.966427] Code: c0 75 1e 8b 73 0c 85 f6 7e 21 48 c1 ea 07 f0 0f b1 53 08 85 c0 75 e5 48 83 c4 18 48 89 c8 5b 5d c3 c1 e0 07 48 8d 5c 05 00 eb d1 <0f> 0b 0f 0b 31 c9 eb e4 66 90 66 2e 0f 1f 84 00 00 00 00 00 0f 
[  680.966427] RIP: tdb_htrie_extend_rec+0x7c/0x90 [tempesta_db] RSP: ffff8a58e5403518
```

Coudn't reproduce the  crash on the same short response right away, but seen it multiple times. Anyway h2 clients complain about h2 frame content:
```
nghttp --color -vvvvvvn   https://192.168.122.12/short_headers/hello
[  0.010] Connected
[WARNING] Certificate verification failed: Hostname mismatch
The negotiated protocol: h2
[  0.018] send SETTINGS frame <length=12, flags=0x00, stream_id=0>
          (niv=2)
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):100]
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):65535]
[  0.018] send PRIORITY frame <length=5, flags=0x00, stream_id=3>
          (dep_stream_id=0, weight=201, exclusive=0)
[  0.018] send PRIORITY frame <length=5, flags=0x00, stream_id=5>
          (dep_stream_id=0, weight=101, exclusive=0)
[  0.018] send PRIORITY frame <length=5, flags=0x00, stream_id=7>
          (dep_stream_id=0, weight=1, exclusive=0)
[  0.019] send PRIORITY frame <length=5, flags=0x00, stream_id=9>
          (dep_stream_id=7, weight=1, exclusive=0)
[  0.019] send PRIORITY frame <length=5, flags=0x00, stream_id=11>
          (dep_stream_id=3, weight=1, exclusive=0)
[  0.019] send HEADERS frame <length=53, flags=0x25, stream_id=13>
          ; END_STREAM | END_HEADERS | PRIORITY
          (padlen=0, dep_stream_id=11, weight=16, exclusive=0)
          ; Open new stream
          :method: GET
          :path: /short_headers/hello
          :scheme: https
          :authority: 192.168.122.12
          accept: */*
          accept-encoding: gzip, deflate
          user-agent: nghttp2/1.40.0
[  0.019] recv SETTINGS frame <length=6, flags=0x00, stream_id=0>
          (niv=1)
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):2147483647]
[  0.019] recv WINDOW_UPDATE frame <length=4, flags=0x00, stream_id=0>
          (window_size_increment=2147418112)
[  0.019] recv SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.019] send SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.019] recv (stream_id=13) :status: 200
[  0.019] recv (stream_id=13) date: Wed, 13 May 2020 11:48:51 GMT
[  0.019] recv (stream_id=13) content-type: application/octet-stream
[  0.019] recv (stream_id=13) content-length: 14
[  0.019] recv (stream_id=13) last-modified: Wed, 13 May 2020 10:23:51 GMT
[  0.019] recv (stream_id=13) etag: "5ebbcab7-e"
[  0.019] recv (stream_id=13) via: 1.1 distant_server
[  0.019] [ERROR] Invalid HTTP header field was received: frame type: 1, stream: 13, name: [X-Extra-Data1], value: [q]
[  0.019] [INVALID; error=Invalid HTTP header field was received] recv HEADERS frame <length=225, flags=0x04, stream_id=13>
          ; END_HEADERS
          (padlen=0)
          ; First response header
[  0.019] send RST_STREAM frame <length=4, flags=0x00, stream_id=13>
          (error_code=PROTOCOL_ERROR(0x01))
[  0.019] send GOAWAY frame <length=8, flags=0x00, stream_id=0>
          (last_stream_id=0, error_code=NO_ERROR(0x00), opaque_data(0)=[])
```

Crash on bigger responses constructed the same way:
```
------------[ cut here ]------------
[ 2600.627354] kernel BUG at /home/user/qtc/release/tempesta/tempesta_fw/hpack.c:2646!
[ 2600.629827] invalid opcode: 0000 [#1] SMP PTI
[ 2600.630476] Modules linked in: tempesta_fw(O) tempesta_db(O) tempesta_tls(O) tempesta_lib(O) sha256_ssse3 sha512_ssse3 sha512_generic ccm kvm_intel iTCO_wdt iTCO_vendor_support binfmt_misc kvm irqbypass crct10dif_pclmul crc32_pclmul snd_hda_codec_generic snd_hda_intel ghash_clmulni_intel snd_hda_codec sg snd_hda_core snd_hwdep virtio_balloon snd_pcm snd_timer snd soundcore virtio_console pcspkr serio_raw virtio_gpu ttm lpc_ich mfd_core evdev drm_kms_helper drm shpchp button ip_tables x_tables autofs4 ext4 crc16 mbcache jbd2 fscrypto ecb aesni_intel crypto_simd cryptd glue_helper aes_x86_64 crc32c_generic sr_mod cdrom virtio_blk virtio_net crc32c_intel psmouse ahci libahci sym53c8xx libata i2c_i801 scsi_transport_spi ehci_pci uhci_hcd ehci_hcd scsi_mod virtio_pci usbcore virtio_ring virtio usb_common
[ 2600.633799]  [last unloaded: tempesta_lib]
[ 2600.633799] CPU: 0 PID: 555 Comm: nginx Tainted: G           O    4.14.0-tempesta-kmemleak-amd64 #1 Debian 4.14.32-tfw7-1
[ 2600.633799] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS ?-20191223_100556-anatol 04/01/2014
[ 2600.633799] task: ffff9b017c103c00 task.stack: ffffbc7540b58000
[ 2600.633799] RIP: 0010:tfw_hpack_rbtree_erase+0x67b/0x700 [tempesta_fw]
[ 2600.633799] RSP: 0018:ffff9b01a5403758 EFLAGS: 00010286
[ 2600.633799] RAX: 0000000000000070 RBX: ffff9b0152bb0058 RCX: 0000000000000038
[ 2600.633799] RDX: ffff9b0152bb0058 RSI: 0000000000000070 RDI: ffff9b017c17b1a8
[ 2600.633799] RBP: ffffffffffffffff R08: ffff9b0152bb0020 R09: ffff9b0152bb0058
[ 2600.633799] R10: 000000000000015f R11: 00000000000001f8 R12: ffff9b017c17b1b0
[ 2600.633799] R13: 0000000000000000 R14: ffff9b017c17b1a8 R15: ffff9b017c17b190
[ 2600.633799] FS:  00007f79e6972740(0000) GS:ffff9b01a5400000(0000) knlGS:0000000000000000
[ 2600.633799] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[ 2600.633799] CR2: 000055d61a4ee000 CR3: 000000007be98005 CR4: 00000000003606f0
[ 2600.633799] Call Trace:
[ 2600.633799]  <IRQ>
[ 2600.633799]  tfw_hpack_add_node+0x24b/0x540 [tempesta_fw]
[ 2600.633799]  tfw_hpack_encode+0x1079/0x12a0 [tempesta_fw]
[ 2600.633799]  tfw_http_resp_cache_cb+0x366/0x610 [tempesta_fw]
[ 2600.633799]  ? netif_rx_internal+0x12a/0x140
[ 2600.633799]  ? tfw_h2_resp_add_loc_hdrs+0xa0/0xa0 [tempesta_fw]
[ 2600.633799]  tfw_cache_process+0x72/0x2a0 [tempesta_fw]
[ 2600.633799]  ? tfw_gfsm_move+0x14b/0x180 [tempesta_fw]
[ 2600.633799]  ? kvm_clock_get_cycles+0x1a/0x20
[ 2600.633799]  tfw_http_resp_cache+0xe1/0x150 [tempesta_fw]
[ 2600.633799]  tfw_http_msg_process_generic+0x893/0xd10 [tempesta_fw]
[ 2600.633799]  ? __dev_queue_xmit+0x2ec/0x7d0
[ 2600.633799]  ? pg_skb_alloc+0x253/0x470
[ 2600.633799]  tfw_http_msg_process+0x84/0xc0 [tempesta_fw]
[ 2600.633799]  __gfsm_fsm_exec+0x56/0x90 [tempesta_fw]
[ 2600.633799]  ? consume_skb+0x27/0x80
[ 2600.633799]  tfw_connection_recv+0x41/0x60 [tempesta_fw]
[ 2600.633799]  ? tfw_connection_send+0x30/0x30 [tempesta_fw]
[ 2600.633799]  ss_tcp_process_data+0x1ea/0x480 [tempesta_fw]
[ 2600.633799]  ? tcp_transmit_skb+0x535/0x9c0
[ 2600.633799]  ss_tcp_data_ready+0x43/0x90 [tempesta_fw]
[ 2600.633799]  tcp_rcv_established+0x4d2/0x570
[ 2600.633799]  tcp_v4_do_rcv+0x129/0x1d0
[ 2600.633799]  tcp_v4_rcv+0x947/0xa50
[ 2600.633799]  ip_local_deliver_finish+0x9a/0x1c0
[ 2600.633799]  ip_local_deliver+0x6b/0xe0
[ 2600.633799]  ? nf_hook_slow+0x43/0xc0
[ 2600.633799]  ip_rcv+0x289/0x3c0
[ 2600.633799]  ? inet_del_offload+0x40/0x40
[ 2600.633799]  __netif_receive_skb_core+0x84f/0xb30
[ 2600.633799]  ? blk_mq_free_request+0xfe/0x110
[ 2600.633799]  ? process_backlog+0xa3/0x160
[ 2600.633799]  process_backlog+0xa3/0x160
[ 2600.633799]  net_rx_action+0x28e/0x3f0
[ 2600.633799]  __do_softirq+0x10f/0x2a8
[ 2600.633799]  do_softirq_own_stack+0x2a/0x40
[ 2600.633799]  </IRQ>
[ 2600.633799]  do_softirq.part.17+0x3d/0x50
[ 2600.633799]  __local_bh_enable_ip+0x55/0x60
[ 2600.633799]  ip_finish_output2+0x1b3/0x3b0
[ 2600.633799]  ? ip_output+0x71/0xe0
[ 2600.633799]  ip_output+0x71/0xe0
[ 2600.633799]  ? ip_queue_xmit+0x5c/0x3b0
[ 2600.633799]  tcp_transmit_skb+0x535/0x9c0
[ 2600.633799]  tcp_write_xmit+0x223/0x11c0
[ 2600.633799]  ? _copy_from_iter_full+0x93/0x230
[ 2600.633799]  __tcp_push_pending_frames+0x2d/0xd0
[ 2600.633799]  tcp_sendmsg_locked+0x68e/0xe60
[ 2600.633799]  tcp_sendmsg+0x27/0x40
[ 2600.633799]  sock_sendmsg+0x36/0x40
[ 2600.633799]  sock_write_iter+0x87/0x100
[ 2600.633799]  do_iter_readv_writev+0x11b/0x190
[ 2600.633799]  do_iter_write+0x80/0x190
[ 2600.633799]  vfs_writev+0x98/0x110
[ 2600.633799]  ? do_writev+0x5e/0xf0
[ 2600.633799]  do_writev+0x5e/0xf0
[ 2600.633799]  ? SyS_pread64+0x90/0xb0
[ 2600.633799]  do_syscall_64+0x8d/0x120
[ 2600.633799]  entry_SYSCALL_64_after_hwframe+0x3d/0xa2
[ 2600.633799] RIP: 0033:0x7f79e52aafe4
[ 2600.633799] RSP: 002b:00007fff44257aa8 EFLAGS: 00000246 ORIG_RAX: 0000000000000014
[ 2600.633799] RAX: ffffffffffffffda RBX: 00007f79e69462c8 RCX: 00007f79e52aafe4
[ 2600.633799] RDX: 0000000000000001 RSI: 00007fff44257b60 RDI: 000000000000000b
[ 2600.633799] RBP: 000055d61a4ece10 R08: 000055d61a4ecdf0 R09: 00007fff44257b60
[ 2600.633799] R10: 000055d61a4ced50 R11: 0000000000000246 R12: 00007fff44257b40
[ 2600.633799] R13: 000055d6198cd48c R14: 000000007fffefff R15: 000000000000000e
[ 2600.633799] Code: 48 29 c8 4d 85 c0 48 c7 c1 ff ff ff ff 48 0f 44 c1 66 89 43 0c e9 9d fa ff ff 0f 0b 41 b9 ff ff ff ff 31 c9 e9 25 fb ff ff 0f 0b <0f> 0b 48 85 d2 74 59 41 89 d1 48 89 d1 45 29 d1 66 44 89 48 0c 
[ 2600.633799] RIP: tfw_hpack_rbtree_erase+0x67b/0x700 [tempesta_fw] RSP: ffff9b01a5403758
```

--- a/tempesta_fw/cache.c
+++ b/tempesta_fw/cache.c
@@ -1152,23 +1152,22 @@ tfw_cache_h2_copy_hdr(TfwCacheEntry *ce, char **p, TdbVRec **trec, TfwStr *hdr,
 		return -ENOMEM;
 	}
 
-	CSTR_MOVE_HDR();
-
 	if (TFW_STR_DUP(hdr)) {
-		CSTR_WRITE_HDR(TFW_STR_DUPLICATE, hdr->nchunks);
 		CSTR_MOVE_HDR();
+		CSTR_WRITE_HDR(TFW_STR_DUPLICATE, hdr->nchunks);
 	}
 
 	TFW_STR_FOR_EACH_DUP(dup, hdr, dup_end) {
-		unsigned int prev_len = ce->hdr_len;
+		unsigned int prev_len;
 
 		if (dupl) {
 			TFW_STR_INIT(&s_nm);
 			TFW_STR_INIT(&s_val);
 			tfw_http_hdr_split(dup, &s_nm, &s_val, true);
 			st_index = dup->hpack_idx;
-			CSTR_MOVE_HDR();
 		}
+		CSTR_MOVE_HDR();
+		prev_len = ce->hdr_len;
 
 		if (st_index) {
 			if (tfw_cache_h2_copy_int(&ce->hdr_len, st_index, 0xF,

--- a/tempesta_fw/cache.c
+++ b/tempesta_fw/cache.c
@@ -723,8 +723,8 @@ tfw_cache_build_resp_hdr(TDB *db, TfwHttpResp *resp, TfwHdrMods *hmods,
 		r = write_actor(db, trec, resp, p, s->len, &dc_iter);
 		if (unlikely(r))
 			break;
-		*acc_len += dc_iter.acc_len;;
 	}
+	*acc_len += dc_iter.acc_len;
 
 	return r;
 }

--- a/tempesta_fw/hpack.c
+++ b/tempesta_fw/hpack.c
@@ -2594,15 +2594,14 @@ tfw_hpack_rbtree_ins_rebalance(TfwHPackETbl *__restrict tbl,
 static void
 tfw_hpack_rbtree_del_rebalance(TfwHPackETbl *__restrict tbl,
 			       TfwHPackNode *__restrict nchild,
-			       TfwHPackNode *__restrict parent,
-			       bool left)
+			       TfwHPackNode *__restrict parent)
 {
 	BUG_ON(!tbl->root);
 
 	while (!nchild || (nchild != tbl->root && HPACK_RB_IS_BLACK(nchild))) {
 		TfwHPackNode *brother, *l_neph, *r_neph;
 
-		if (left) {
+		if (nchild == HPACK_NODE_COND(tbl, parent->left)) {
 			BUG_ON(HPACK_NODE_EMPTY(parent->right));
 			brother = HPACK_NODE(tbl, parent->right);
 			if (HPACK_RB_IS_RED(brother)) {
@@ -2706,12 +2705,11 @@ tfw_hpack_rbtree_min(TfwHPackETbl *__restrict tbl,
 /*
  * Procedure for branches replacement in red-black tree.
  */
-static inline bool
+static inline void
 tfw_hpack_rbtree_replace(TfwHPackETbl *__restrict tbl,
 			 TfwHPackNode *__restrict old,
 			 TfwHPackNode *__restrict new)
 {
-	bool left = true;
 	TfwHPackNode *parent = HPACK_NODE_COND(tbl, old->parent);
 
 	BUG_ON(!old);
@@ -2730,13 +2728,10 @@ tfw_hpack_rbtree_replace(TfwHPackETbl *__restrict tbl,
 		WARN_ON_ONCE(HPACK_NODE_EMPTY(parent->right)
 			     || old != HPACK_NODE(tbl, parent->right));
 		parent->right = HPACK_NODE_COND_OFF(tbl, new);
-		left = false;
 	}
 
 	if (new)
 		new->parent = HPACK_NODE_COND_OFF(tbl, parent);
-
-	return left;
 }
 
 /*
@@ -2834,15 +2829,15 @@ tfw_hpack_rbtree_erase(TfwHPackETbl *__restrict tbl,
 {
 	TfwHPackNode *nchild, *sv = node;
 	TfwHPackNode *parent = HPACK_NODE_COND(tbl, node->parent);
-	bool left_child = false, sv_black = HPACK_RB_IS_BLACK(sv);
+	bool sv_black = HPACK_RB_IS_BLACK(sv);
 
 	if (HPACK_NODE_EMPTY(node->left)) {
 		nchild = HPACK_NODE_COND(tbl, node->right);
-		left_child = tfw_hpack_rbtree_replace(tbl, node, nchild);
+		tfw_hpack_rbtree_replace(tbl, node, nchild);
 	}
 	else if (HPACK_NODE_EMPTY(node->right)) {
 		nchild = HPACK_NODE_COND(tbl, node->left);
-		left_child = tfw_hpack_rbtree_replace(tbl, node, nchild);
+		tfw_hpack_rbtree_replace(tbl, node, nchild);
 	}
 	else {
 		TfwHPackNode *n_left;
@@ -2856,7 +2851,7 @@ tfw_hpack_rbtree_erase(TfwHPackETbl *__restrict tbl,
 			TfwHPackNode *n_right;
 
 			parent = HPACK_NODE(tbl, sv->parent);
-			left_child = tfw_hpack_rbtree_replace(tbl, sv, nchild);
+			tfw_hpack_rbtree_replace(tbl, sv, nchild);
 
 			n_right = HPACK_NODE(tbl, node->right);
 			n_right->parent = HPACK_NODE_OFF(tbl, sv);
@@ -2881,7 +2876,7 @@ tfw_hpack_rbtree_erase(TfwHPackETbl *__restrict tbl,
 	 * the last).
 	 */
 	if (sv_black && tbl->root)
-		tfw_hpack_rbtree_del_rebalance(tbl, nchild, parent, left_child);
+		tfw_hpack_rbtree_del_rebalance(tbl, nchild, parent);
 }
 
 static inline void

--- a/tempesta_fw/hpack.c
+++ b/tempesta_fw/hpack.c
@@ -2538,6 +2538,7 @@ tfw_hpack_rbtree_ins_rebalance(TfwHPackETbl *__restrict tbl,
 				HPACK_RB_SET_BLACK(uncle);
 				HPACK_RB_SET_RED(gparent);
 				parent = HPACK_NODE_COND(tbl, gparent->parent);
+				new = gparent;
 			}
 			else {
 				if (!HPACK_NODE_EMPTY(parent->right)
@@ -2545,6 +2546,11 @@ tfw_hpack_rbtree_ins_rebalance(TfwHPackETbl *__restrict tbl,
 				{
 					tfw_hpack_rbtree_left_rt(tbl, parent);
 					parent = new;
+					/*
+					 * Don't need `new = HPACK_NODE_COND(tbl, parent->left);`.
+					 * This is the last iteration, because the parent will turn
+					 * black below
+					 */
 				}
 				HPACK_RB_SET_BLACK(parent);
 				HPACK_RB_SET_RED(gparent);
@@ -2558,6 +2564,7 @@ tfw_hpack_rbtree_ins_rebalance(TfwHPackETbl *__restrict tbl,
 				HPACK_RB_SET_BLACK(uncle);
 				HPACK_RB_SET_RED(gparent);
 				parent = HPACK_NODE_COND(tbl, gparent->parent);
+				new = gparent;
 			}
 			else {
 				if (!HPACK_NODE_EMPTY(parent->left)
@@ -2565,6 +2572,11 @@ tfw_hpack_rbtree_ins_rebalance(TfwHPackETbl *__restrict tbl,
 				{
 					tfw_hpack_rbtree_right_rt(tbl, parent);
 					parent = new;
+					/*
+					 * Don't need `new = HPACK_NODE_COND(tbl, parent->right);`.
+					 * This is the last iteration, because the parent will turn
+					 * black below
+					 */
 				}
 				HPACK_RB_SET_BLACK(parent);
 				HPACK_RB_SET_RED(gparent);

--- a/tempesta_fw/hpack.c
+++ b/tempesta_fw/hpack.c
@@ -2628,6 +2628,7 @@ tfw_hpack_rbtree_del_rebalance(TfwHPackETbl *__restrict tbl,
 			{
 				HPACK_RB_SET_RED(brother);
 				nchild = parent;
+				parent = HPACK_NODE_COND(tbl, nchild->parent);
 			}
 			else
 			{
@@ -2665,6 +2666,7 @@ tfw_hpack_rbtree_del_rebalance(TfwHPackETbl *__restrict tbl,
 			{
 				HPACK_RB_SET_RED(brother);
 				nchild = parent;
+				parent = HPACK_NODE_COND(tbl, nchild->parent);
 			}
 			else
 			{

--- a/tempesta_fw/hpack.c
+++ b/tempesta_fw/hpack.c
@@ -2954,11 +2954,15 @@ tfw_hpack_rbuf_calc(TfwHPackETbl *__restrict tbl, unsigned short new_size,
 
 static inline void
 tfw_hpack_rbuf_commit(TfwHPackETbl *__restrict tbl,
+		      TfwStr *__restrict hdr,
 		      TfwHPackNode *__restrict del_list[],
 		      TfwHPackNodeIter *__restrict place,
 		      TfwHPackETblIter *__restrict iter)
 {
 	int i;
+	bool was_del = false;
+	const TfwHPackNode *node = NULL;
+	TfwHPackETblRes res = HPACK_IDX_ST_NOT_FOUND;
 
 	for (i = 0; i < HPACK_MAX_ENC_EVICTION; ++i) {
 		TfwHPackNode *del_node = del_list[i];
@@ -2966,6 +2970,17 @@ tfw_hpack_rbuf_commit(TfwHPackETbl *__restrict tbl,
 		if (!del_node)
 			break;
 		tfw_hpack_rbtree_erase(tbl, del_node);
+		was_del = true;
+	}
+
+	/*
+	 * If there was a deletion, the place may turn out to be invalid as a result
+	 * of reducing the procedure for deleting a node with two children to
+	 * deleting a node with less than two children.
+	 */
+	if (was_del) {
+		res = tfw_hpack_rbtree_find(tbl, hdr, &node, place);
+		WARN_ON_ONCE(res == HPACK_IDX_ST_FOUND);
 	}
 
 	tfw_hpack_rbtree_add(tbl, iter->last, place);
@@ -3093,7 +3108,7 @@ commit:
 	ptr = tfw_hpack_write(&s_nm, it.last->hdr);
 	tfw_hpack_write(&s_val, ptr);
 
-	tfw_hpack_rbuf_commit(tbl, del_list, place, &it);
+	tfw_hpack_rbuf_commit(tbl, hdr, del_list, place, &it);
 
 	WARN_ON_ONCE(tbl->rb_len > tbl->size);
 

--- a/tempesta_fw/t/unit/test_hpack.c
+++ b/tempesta_fw/t/unit/test_hpack.c
@@ -1901,6 +1901,19 @@ TEST(hpack, enc_table_rbtree)
 #undef HDR_VALUE_5
 }
 
+#define ADD_NODE(s, n)								\
+do {										\
+	res = tfw_hpack_rbtree_find(tbl, s, &n, &pl);				\
+	EXPECT_EQ(res, HPACK_IDX_ST_NOT_FOUND);					\
+	EXPECT_NULL(n);								\
+	EXPECT_OK(tfw_hpack_add_node(tbl, s, &pl, TFW_H2_TRANS_INPLACE));	\
+	bzero_fast(&pl, sizeof(pl));						\
+	res = tfw_hpack_rbtree_find(tbl, s, &n, &pl);				\
+	EXPECT_EQ(res, HPACK_IDX_ST_FOUND);					\
+	EXPECT_NULL(pl.parent);							\
+	EXPECT_NOT_NULL(n);							\
+} while (0)
+
 TEST(hpack, rbtree_ins_rebalance)
 {
 	TfwHPackETbl *tbl;
@@ -1954,19 +1967,6 @@ TEST(hpack, rbtree_ins_rebalance)
 	collect_compound_str(s10, s10_value, 0);
 
 	tbl = &ctx.hpack.enc_tbl;
-
-#define ADD_NODE(s, n)								\
-do {										\
-	res = tfw_hpack_rbtree_find(tbl, s, &n, &pl);				\
-	EXPECT_EQ(res, HPACK_IDX_ST_NOT_FOUND);					\
-	EXPECT_NULL(n);								\
-	EXPECT_OK(tfw_hpack_add_node(tbl, s, &pl, TFW_H2_TRANS_INPLACE));	\
-	bzero_fast(&pl, sizeof(pl));						\
-	res = tfw_hpack_rbtree_find(tbl, s, &n, &pl);				\
-	EXPECT_EQ(res, HPACK_IDX_ST_FOUND);					\
-	EXPECT_NULL(pl.parent);							\
-	EXPECT_NOT_NULL(n);							\
-} while (0)
 
 	ADD_NODE(s1, n1);
 	ADD_NODE(s2, n2);
@@ -2068,8 +2068,6 @@ do {										\
 	EXPECT_EQ(HPACK_NODE_COND(tbl, n2->right), n1);
 	EXPECT_NULL(HPACK_NODE_COND(tbl, n1->left));
 	EXPECT_NULL(HPACK_NODE_COND(tbl, n1->right));
-
-#undef ADD_NODE
 }
 
 TEST(hpack, rbtree_del_rebalance)
@@ -2125,19 +2123,6 @@ TEST(hpack, rbtree_del_rebalance)
 	collect_compound_str(s10, s10_value, 0);
 
 	tbl = &ctx.hpack.enc_tbl;
-
-#define ADD_NODE(s, n)								\
-do {										\
-	res = tfw_hpack_rbtree_find(tbl, s, &n, &pl);				\
-	EXPECT_EQ(res, HPACK_IDX_ST_NOT_FOUND);					\
-	EXPECT_NULL(n);								\
-	EXPECT_OK(tfw_hpack_add_node(tbl, s, &pl, TFW_H2_TRANS_INPLACE));	\
-	bzero_fast(&pl, sizeof(pl));						\
-	res = tfw_hpack_rbtree_find(tbl, s, &n, &pl);				\
-	EXPECT_EQ(res, HPACK_IDX_ST_FOUND);					\
-	EXPECT_NULL(pl.parent);							\
-	EXPECT_NOT_NULL(n);							\
-} while (0)
 
 	ADD_NODE(s1, n1);
 	ADD_NODE(s2, n2);
@@ -2240,9 +2225,9 @@ do {										\
 	EXPECT_EQ(HPACK_NODE_COND(tbl, n9->right), n10);
 	EXPECT_NULL(HPACK_NODE_COND(tbl, n10->left));
 	EXPECT_NULL(HPACK_NODE_COND(tbl, n10->right));
+}
 
 #undef ADD_NODE
-}
 
 TEST_SUITE(hpack)
 {

--- a/tempesta_fw/t/unit/test_hpack.c
+++ b/tempesta_fw/t/unit/test_hpack.c
@@ -2072,6 +2072,178 @@ do {										\
 #undef ADD_NODE
 }
 
+TEST(hpack, rbtree_del_rebalance)
+{
+	TfwHPackETbl *tbl;
+	TfwHPackETblRes res;
+	TfwHPackNodeIter pl = {};
+	const TfwHPackNode *n1 = NULL, *n2 = NULL, *n3 = NULL, *n4 = NULL;
+	const TfwHPackNode *n5 = NULL, *n6 = NULL, *n7 = NULL, *n8 = NULL;
+	const TfwHPackNode *n9 = NULL, *n10 = NULL;
+
+	TFW_STR(col, ":");
+	TFW_STR(s1, "1");
+	TFW_STR(s1_value, "1");
+	TFW_STR(s2, "2");
+	TFW_STR(s2_value, "2");
+	TFW_STR(s3, "3");
+	TFW_STR(s3_value, "3");
+	TFW_STR(s4, "4");
+	TFW_STR(s4_value, "4");
+	TFW_STR(s5, "5");
+	TFW_STR(s5_value, "5");
+	TFW_STR(s6, "6");
+	TFW_STR(s6_value, "6");
+	TFW_STR(s7, "7");
+	TFW_STR(s7_value, "7");
+	TFW_STR(s8, "8");
+	TFW_STR(s8_value, "8");
+	TFW_STR(s9, "9");
+	TFW_STR(s9_value, "9");
+	TFW_STR(s10, "a");
+	TFW_STR(s10_value, "a");
+
+	collect_compound_str(s1, col, 0);
+	collect_compound_str(s1, s1_value, 0);
+	collect_compound_str(s2, col, 0);
+	collect_compound_str(s2, s2_value, 0);
+	collect_compound_str(s3, col, 0);
+	collect_compound_str(s3, s3_value, 0);
+	collect_compound_str(s4, col, 0);
+	collect_compound_str(s4, s4_value, 0);
+	collect_compound_str(s5, col, 0);
+	collect_compound_str(s5, s5_value, 0);
+	collect_compound_str(s6, col, 0);
+	collect_compound_str(s6, s6_value, 0);
+	collect_compound_str(s7, col, 0);
+	collect_compound_str(s7, s7_value, 0);
+	collect_compound_str(s8, col, 0);
+	collect_compound_str(s8, s8_value, 0);
+	collect_compound_str(s9, col, 0);
+	collect_compound_str(s9, s9_value, 0);
+	collect_compound_str(s10, col, 0);
+	collect_compound_str(s10, s10_value, 0);
+
+	tbl = &ctx.hpack.enc_tbl;
+
+#define ADD_NODE(s, n)								\
+do {										\
+	res = tfw_hpack_rbtree_find(tbl, s, &n, &pl);				\
+	EXPECT_EQ(res, HPACK_IDX_ST_NOT_FOUND);					\
+	EXPECT_NULL(n);								\
+	EXPECT_OK(tfw_hpack_add_node(tbl, s, &pl, TFW_H2_TRANS_INPLACE));	\
+	bzero_fast(&pl, sizeof(pl));						\
+	res = tfw_hpack_rbtree_find(tbl, s, &n, &pl);				\
+	EXPECT_EQ(res, HPACK_IDX_ST_FOUND);					\
+	EXPECT_NULL(pl.parent);							\
+	EXPECT_NOT_NULL(n);							\
+} while (0)
+
+	ADD_NODE(s1, n1);
+	ADD_NODE(s2, n2);
+	ADD_NODE(s3, n3);
+	ADD_NODE(s4, n4);
+	ADD_NODE(s5, n5);
+	ADD_NODE(s6, n6);
+	ADD_NODE(s7, n7);
+	ADD_NODE(s8, n8);
+	ADD_NODE(s9, n9);
+	ADD_NODE(s10, n10);
+
+	/*
+	 *                  n4
+	 *                (BLACK)
+	 *               /      \
+	 *              n2        n6
+	 *            (BLACK)    (BLACK)
+	 *            /    \     /      \
+	 *           n1     n3   n5      n8
+	 *        (BLACK) (BLACK)(BLACK)(RED)
+	 *                               /   \
+	 *                              n7    n9
+	 *                           (BLACK)(BLACK)
+	 *                                        \
+	 *                                         n10
+	 *                                        (RED)
+	 */
+
+	tfw_hpack_rbtree_erase(tbl, (TfwHPackNode *)n1);
+	/*
+	 * Erase the node n1.
+	 *
+	 *                   n4
+	 *                 (BLACK)
+	 *               /        \
+	 *              n2         n6
+	 *            (BLACK)      (BLACK)
+	 *                 \      /      \
+	 *                  n3    n5      n8
+	 *               (BLACK)(BLACK)  (RED)
+	 *                               /   \
+	 *                              n7    n9
+	 *                           (BLACK)(BLACK)
+	 *                                        \
+	 *                                         n10
+	 *                                        (RED)
+	 *
+	 * Since there is no red brother and the brother n3 has no children, then we
+	 * make the brother red.
+	 *
+	 *                  n4
+	 *                (BLACK)
+	 *               /       \
+	 *              n2        n6
+	 *            (BLACK)    (BLACK)
+	 *                 \     /      \
+	 *                  n3   n5      n8
+	 *                (RED)(BLACK)   (RED)
+	 *                               /   \
+	 *                              n7    n9
+	 *                           (BLACK)(BLACK)
+	 *                                        \
+	 *                                         n10
+	 *                                        (RED)
+	 *
+	 * The current node n2 is the parent of erased n1.
+	 * Since there is no red brother, the brother n6 has a red right child n8,
+	 * but the current node n2 is a left child, then we assign the color of the
+	 * parent n4 to the brother n6, make the parent n4 black, make the right
+	 * child of the brother n8 black and make a left turn relative to the parent
+	 * n4.
+	 *
+	 *                           n6
+	 *                         (BLACK)
+	 *                        /       \
+	 *                       n4        n8
+	 *                     (BLACK)    (BLACK)
+	 *                       /  \       /   \
+	 *                      n2   n5     n7    n9
+	 *                   (BLACK)(BLACK)(BLACK)(BLACK)
+	 *                       \                  \
+	 *                        n3                n10
+	 *                       (RED)             (RED)
+	 */
+
+	EXPECT_EQ(HPACK_NODE_COND(tbl, n6->left), n4);
+	EXPECT_EQ(HPACK_NODE_COND(tbl, n4->left), n2);
+	EXPECT_NULL(HPACK_NODE_COND(tbl, n2->left));
+	EXPECT_EQ(HPACK_NODE_COND(tbl, n2->right), n3);
+	EXPECT_EQ(HPACK_NODE_COND(tbl, n4->right), n5);
+	EXPECT_NULL(HPACK_NODE_COND(tbl, n5->left));
+	EXPECT_NULL(HPACK_NODE_COND(tbl, n5->right));
+	EXPECT_EQ(HPACK_NODE_COND(tbl, n6->right), n8);
+	EXPECT_EQ(HPACK_NODE_COND(tbl, n8->left), n7);
+	EXPECT_NULL(HPACK_NODE_COND(tbl, n7->left));
+	EXPECT_NULL(HPACK_NODE_COND(tbl, n7->right));
+	EXPECT_EQ(HPACK_NODE_COND(tbl, n8->right), n9);
+	EXPECT_NULL(HPACK_NODE_COND(tbl, n9->left));
+	EXPECT_EQ(HPACK_NODE_COND(tbl, n9->right), n10);
+	EXPECT_NULL(HPACK_NODE_COND(tbl, n10->left));
+	EXPECT_NULL(HPACK_NODE_COND(tbl, n10->right));
+
+#undef ADD_NODE
+}
+
 TEST_SUITE(hpack)
 {
 	TEST_SETUP(test_h2_setup);
@@ -2089,4 +2261,5 @@ TEST_SUITE(hpack)
 	TEST_RUN(hpack, enc_table_index);
 	TEST_RUN(hpack, enc_table_rbtree);
 	TEST_RUN(hpack, rbtree_ins_rebalance);
+	TEST_RUN(hpack, rbtree_del_rebalance);
 }

--- a/tempesta_fw/t/unit/test_hpack.c
+++ b/tempesta_fw/t/unit/test_hpack.c
@@ -1901,6 +1901,177 @@ TEST(hpack, enc_table_rbtree)
 #undef HDR_VALUE_5
 }
 
+TEST(hpack, rbtree_ins_rebalance)
+{
+	TfwHPackETbl *tbl;
+	TfwHPackETblRes res;
+	TfwHPackNodeIter pl = {};
+	const TfwHPackNode *n1 = NULL, *n2 = NULL, *n3 = NULL, *n4 = NULL;
+	const TfwHPackNode *n5 = NULL, *n6 = NULL, *n7 = NULL, *n8 = NULL;
+	const TfwHPackNode *n9 = NULL, *n10 = NULL;
+
+	TFW_STR(col, ":");
+	TFW_STR(s1, "80");
+	TFW_STR(s1_value, "80");
+	TFW_STR(s2, "70");
+	TFW_STR(s2_value, "70");
+	TFW_STR(s3, "60");
+	TFW_STR(s3_value, "60");
+	TFW_STR(s4, "50");
+	TFW_STR(s4_value, "50");
+	TFW_STR(s5, "40");
+	TFW_STR(s5_value, "40");
+	TFW_STR(s6, "30");
+	TFW_STR(s6_value, "30");
+	TFW_STR(s7, "20");
+	TFW_STR(s7_value, "20");
+	TFW_STR(s8, "55");
+	TFW_STR(s8_value, "55");
+	TFW_STR(s9, "65");
+	TFW_STR(s9_value, "65");
+	TFW_STR(s10, "67");
+	TFW_STR(s10_value, "67");
+
+	collect_compound_str(s1, col, 0);
+	collect_compound_str(s1, s1_value, 0);
+	collect_compound_str(s2, col, 0);
+	collect_compound_str(s2, s2_value, 0);
+	collect_compound_str(s3, col, 0);
+	collect_compound_str(s3, s3_value, 0);
+	collect_compound_str(s4, col, 0);
+	collect_compound_str(s4, s4_value, 0);
+	collect_compound_str(s5, col, 0);
+	collect_compound_str(s5, s5_value, 0);
+	collect_compound_str(s6, col, 0);
+	collect_compound_str(s6, s6_value, 0);
+	collect_compound_str(s7, col, 0);
+	collect_compound_str(s7, s7_value, 0);
+	collect_compound_str(s8, col, 0);
+	collect_compound_str(s8, s8_value, 0);
+	collect_compound_str(s9, col, 0);
+	collect_compound_str(s9, s9_value, 0);
+	collect_compound_str(s10, col, 0);
+	collect_compound_str(s10, s10_value, 0);
+
+	tbl = &ctx.hpack.enc_tbl;
+
+#define ADD_NODE(s, n)								\
+do {										\
+	res = tfw_hpack_rbtree_find(tbl, s, &n, &pl);				\
+	EXPECT_EQ(res, HPACK_IDX_ST_NOT_FOUND);					\
+	EXPECT_NULL(n);								\
+	EXPECT_OK(tfw_hpack_add_node(tbl, s, &pl, TFW_H2_TRANS_INPLACE));	\
+	bzero_fast(&pl, sizeof(pl));						\
+	res = tfw_hpack_rbtree_find(tbl, s, &n, &pl);				\
+	EXPECT_EQ(res, HPACK_IDX_ST_FOUND);					\
+	EXPECT_NULL(pl.parent);							\
+	EXPECT_NOT_NULL(n);							\
+} while (0)
+
+	ADD_NODE(s1, n1);
+	ADD_NODE(s2, n2);
+	ADD_NODE(s3, n3);
+	ADD_NODE(s4, n4);
+	ADD_NODE(s5, n5);
+	ADD_NODE(s6, n6);
+	ADD_NODE(s7, n7);
+	ADD_NODE(s8, n8);
+	ADD_NODE(s9, n9);
+
+	/*
+	 *                 n2
+	 *              (BLACK)
+	 *               /    \
+	 *              n4     n1
+	 *            (RED)  (BLACK)
+	 *            /    \
+	 *           n6     n3
+	 *        (BLACK) (BLACK)
+	 *         /  \     /  \
+	 *        n7  n5   n8   n9
+	 *     (RED)(RED) (RED)(RED)
+	 */
+
+	ADD_NODE(s10, n10);
+
+	/*
+	 * Add the node n10 as the right child of n9.
+	 * Because parent n9 - red and uncle n8 - red, we make them (n9 and n8)
+	 * black, and grandparent n3 - red.
+	 *                 n2
+	 *              (BLACK)
+	 *               /    \
+	 *              n4     n1
+	 *             (RED)  (BLACK)
+	 *            /     \
+	 *           n6     n3
+	 *        (BLACK)  (RED)
+	 *         /  \     /   \
+	 *        n7  n5   n8    n9
+	 *     (RED)(RED)(BLACK)(BLACK)
+	 *                         \
+	 *                         n10
+	 *                         (RED)
+	 * The grandparent n3 becomes the new current node.
+	 * Then, since the parent n4 of the current node n3 (great-grandparent of
+	 * the original n10) is red, and its uncle n1 is not red, and the node n3 is
+	 * the right child of the left child - we rotate to the left relative to the
+	 * parent n4.
+	 *                 n2
+	 *              (BLACK)
+	 *               /    \
+	 *              n3   n1
+	 *            (RED) (BLACK)
+	 *            /   \
+	 *           n4    n9
+	 *         (RED)  (BLACK)
+	 *        /     \      \
+	 *       n6     n8     n10
+	 *    (BLACK) (BLACK) (RED)
+	 *    /  \
+	 *   n7  n5
+	 *  (RED)(RED)
+	 * The new current node n4 is the left child of the current node n3.
+	 * Because the current node n4 is the left child of the left child, then we
+	 * make the parent n3 black and the grandparent n2 red and rotate to the
+	 * right relative to its grandparent n3.
+	 *                    n3
+	 *                 (BLACK)
+	 *               /         \
+	 *              n4          n2
+	 *            (RED)         (RED)
+	 *            /   \         /   \
+	 *           n6    n8      n9     n1
+	 *        (BLACK)(BLACK) (BLACK) (BLACK)
+	 *         /  \               \
+	 *        n7  n5              n10
+	 *     (RED)(RED)            (RED)
+	 */
+
+	EXPECT_EQ(HPACK_NODE_COND(tbl, n3->left), n4);
+	EXPECT_EQ(HPACK_NODE_COND(tbl, n4->left), n6);
+	EXPECT_EQ(HPACK_NODE_COND(tbl, n6->left), n7);
+	EXPECT_NULL(HPACK_NODE_COND(tbl, n7->left));
+	EXPECT_NULL(HPACK_NODE_COND(tbl, n7->right));
+	EXPECT_EQ(HPACK_NODE_COND(tbl, n6->right), n5);
+	EXPECT_NULL(HPACK_NODE_COND(tbl, n5->left));
+	EXPECT_NULL(HPACK_NODE_COND(tbl, n5->right));
+	EXPECT_EQ(HPACK_NODE_COND(tbl, n4->right), n8);
+	EXPECT_NULL(HPACK_NODE_COND(tbl, n8->left));
+	EXPECT_NULL(HPACK_NODE_COND(tbl, n8->right));
+	EXPECT_EQ(HPACK_NODE_COND(tbl, n3->right), n2);
+	EXPECT_EQ(HPACK_NODE_COND(tbl, n2->left), n9);
+	EXPECT_NULL(HPACK_NODE_COND(tbl, n9->left));
+	EXPECT_EQ(HPACK_NODE_COND(tbl, n9->right), n10);
+	EXPECT_NULL(HPACK_NODE_COND(tbl, n10->left));
+	EXPECT_NULL(HPACK_NODE_COND(tbl, n10->right));
+	EXPECT_EQ(HPACK_NODE_COND(tbl, n2->right), n1);
+	EXPECT_NULL(HPACK_NODE_COND(tbl, n1->left));
+	EXPECT_NULL(HPACK_NODE_COND(tbl, n1->right));
+
+#undef ADD_NODE
+}
+
 TEST_SUITE(hpack)
 {
 	TEST_SETUP(test_h2_setup);
@@ -1917,4 +2088,5 @@ TEST_SUITE(hpack)
 	TEST_RUN(hpack, enc_table_hdr_write);
 	TEST_RUN(hpack, enc_table_index);
 	TEST_RUN(hpack, enc_table_rbtree);
+	TEST_RUN(hpack, rbtree_ins_rebalance);
 }

